### ISSUE TITLE
Possible issue with modifiedPaths for subdocuments

### ISSUE THREAD

**Issue by JasonCust** [Automattic/mongoose#6134]
**Do you want to request a *feature* or report a *bug*?**
This is similar to a current open bug: #5904

**What is the current behavior?**
Currently, if a subdocument is modified the full path is included in `doc.modifiedPaths()` whereas if the property was modified via replacing a subdocument only the subdocument path is included.

**If the current behavior is a bug, please provide the steps to reproduce.**
```js
const personSchema = new mongoose.Schema({
  name: { type: String },
  colors: {
    primary: {
      type: String,
      default: 'white',
      enum: ['blue', 'green', 'red', 'purple', 'yellow', 'white', 'black']
    }
  }
});

const Person = connection.model('Person', personSchema);

const luke = new Person({
  name: 'Luke',
  colors: {
    primary: 'blue'
  },
});
console.dir(luke.modifiedPaths(), { depth: null, color: true });
// GOOD: [ 'name', 'colors', 'colors.primary' ]

const obiwan = new Person({ name: 'Obi-Wan' });
obiwan.colors.primary = 'blue';
console.dir(obiwan.modifiedPaths(), { depth: null, color: true });
// GOOD: [ 'name', 'colors', 'colors.primary' ]

const anikan = new Person({ name: 'Anikan' });
anikan.colors = { primary: 'blue' };
console.dir(anikan.modifiedPaths(), { depth: null, color: true });
// BAD: [ 'name', 'colors' ]
```

**What is the expected behavior?**
```js
// EXPECTED: [ 'name', 'colors', 'colors.primary' ]
```
**Please mention your node.js, mongoose and MongoDB version.**
Node: 4.8.7, 6.12.2, 8.6.0 (only three tested on)
Mongoose: 4.13.11, 5.0.5 (only two tested on)
MongoDB: 3.4.10, 3.6.0 (only two tested on)

**Comment by vkarpov15** [comment]
Thanks for reporting, will work on a fix but currently behind on issues so please be patient :+1:

**Comment by vkarpov15** [comment]
I'm going to have to put this fix behind a flag (`modifiedPaths({ includeChildren: true })`) because the fix conflicts with the fixes for #4224. I can see both approaches being correct, even though in hindsight I think the right approach is for `modifiedPaths()` to include more paths rather than less.

In general, I think going forward we're going to make a stronger distinction between direct modified paths and the return value from `modifiedPaths()`. The former should really only contain paths that you explicitly set, so `anakin.colors = { primary: blue }` will only contain 'colors' because that's the path that you called `set()` on and that's what we use internally for `save()`, etc. The latter should contain whatever makes sense for clients of mongoose, like your lib :+1:

**Comment by thesirtaylor** [comment]
H

> I'm going to have to put this fix behind a flag (`modifiedPaths({ includeChildren: true })`) because the fix conflicts with the fixes for #4224. I can see both approaches being correct, even though in hindsight I think the right approach is for `modifiedPaths()` to include more paths rather than less.
> 
> In general, I think going forward we're going to make a stronger distinction between direct modified paths and the return value from `modifiedPaths()`. The former should really only contain paths that you explicitly set, so `anakin.colors = { primary: blue }` will only contain 'colors' because that's the path that you called `set()` on and that's what we use internally for `save()`, etc. The latter should contain whatever makes sense for clients of mongoose, like your lib 👍

What if I'm interested in returning the new value of the modified path, how do I go about that?

**Comment by vkarpov15** [comment]
@thesirtaylor you can do something like `doc.modifiedPaths().map(path => doc.get(path))`, does that work?
